modules/wallet/update.go

```diff
@@ -18,10 +18,60 @@ type historicOutput struct {
 	val types.Currency
 }
 
+// isFutureWalletAddress is a helper function that checks if an UnlockHash is
+// derived from one of the wallet's future keys. If it is all future keys up to
+// this one will be added to the spendable keys or a blockhain rescan will be
+// initiated
+func (w *Wallet) isFutureWalletAddress(uh types.UnlockHash) bool {
+	if _, exists := w.futureKeys[uh]; !exists {
+		return false
+	}
+
+	// generate and add spendable keys until the future key is present in w.keys
+	foundKey := false
+	for i := 0; i < 1000; i++ {
+		_, err := w.nextPrimarySeedAddress(w.dbTx)
+		if err != nil {
+			return false
+		}
+
+		if _, exists := w.keys[uh]; exists {
+			foundKey = true
+			break
+		}
+	}
+
+	// if future key was not among the next 1000 keys rescan the blockchain
+	if !foundKey {
```

Comment:
under what circumstances would this happen? I figure that if the key isn't in the lookahead, you don't know that it's one of your keys anyway, so how would you know to rescan the blockchain?

In the extreme case (`InitFromSeed`), we basically just want to use a huge lookahead (think 1 million keys) so that the probability of a rescan is as low as possible. Currently, during `InitFromSeed`, we rescan if the highest index seen in the blockchain was in the higher 50% of the keys we generated. I'm not sure what the rescan rules should be for the lookahead, but it will likely be a similar construction. I think in general we can rely on the assumption that addresses will be used in roughly monotonically-increasing order -- that is, keys with a higher index will appear later in the blockchain, and if they don't, the discrepancy will be no greater than ~1000 keys. 

My thinking is: there's a bit of performance cost associated with a large lookahead, but the cost of a rescan is vastly greater. So we should aim for a lookahead that reduces the probability of a rescan as much as possible without impacting performance too severely. To estimate this, it'd be good to have some benchmark numbers showing how long it takes to generate a new key.

Finally, again wrt `InitFromSeed`, we don't want to have to keep using a huge lookahead forever. Ideally, once we were confident that we had found all the uses of the seed, we would drop the lookahead down to a much lower value (i.e. the default). But when can we be sure that we've found all the uses? Naively, we can say that we've seen all uses once we've done a full scan of a synced blockchain. In other words, we should be able to drop down to the default lookahead as soon as the call to `cs.ConsensusSetSubscribe` completes.